### PR TITLE
Feat/improve tool ergonomics

### PR DESCRIPTION
**Pull request by hojarasca**
Added integration specs
Colors for match or not match in github
Extracted environment creation in its own class to avoid repeated checks
Several fixes found during testing.

Tests are not passing in this branch because for some reason etherscan is rate limiting in a more strict way the request from the CI.

### CODE REVIEW

**Review by hojarasca — COMMENTED** [review]
Round 1 of comments. Round 2 later today :)

File: cli/src/lib/block-explorer-client.ts [hunk 1]
```diff
@@ -95,19 +107,46 @@ export class BlockExplorerClient {
...
+      // We cannot recreate the dir structure in this case. We also don't know the right file name or file type.
       if (e instanceof SyntaxError) {
-        const sources = { "main.sol": { content: rawSourceCode } };
-        const data = new ContractData(result[0].ContractName, sources, contractAddr);
+        const content = { content: result[0].SourceCode };
+        const data = new ContractData(
+          result[0].ContractName,
+          { "contract.sol": content },
+          contractAddr
+        );
```
Comment (hojarasca) [inline cli/src/lib/block-explorer-client.ts:132]: i know this error handling sounds terrible. But I'm 97% sure that this is close to the best we can do in this case.

File: cli/src/lib/cli.ts [hunk 2]
```diff
@@ -24,6 +38,18 @@ export const cli = async () => {
...
+      type: "string",
+      describe: "Ethereum rpc url",
+      demandOption: false,
+    })
+    .middleware((yargs) => {
+      env.withNetwork(NetworkSchema.parse(yargs.network));
+      env.withRpcUrl(yargs.rpcUrl);
+      env.withEtherscanApiKey(yargs.ethscankey);
+      env.withGithubApiKey(yargs.githubApiKey);
+    })
```
Comment (hojarasca) [inline cli/src/lib/cli.ts:52]: This is the middleware to load the configuration into the env object

File: cli/src/lib/cli.ts [hunk 3]
```diff
@@ -40,12 +66,7 @@ export const cli = async () => {
             default: "main",
           }),
       async (yargs) => {
-        await checkCommand(
-          yargs.ethscankey,
-          NetworkSchema.parse(yargs.network),
-          yargs.upgradeDirectory,
-          yargs.ref
-        );
+        await checkCommand(env, yargs.upgradeDirectory, yargs.ref);
```
Comment (hojarasca) [inline cli/src/lib/cli.ts:69]: This is the pattern now. Just send the env object with all the configuration.

File: cli/src/lib/cli.ts [hunk 4]
```diff
@@ -157,18 +166,26 @@ export const cli = async () => {
...
-        );
+        await downloadCode(env, yargs.upgradeDir, yargs.targetSourceCodeDir, filter, yargs.ref);
       }
     )
     .demandCommand(1, "Please specify a command")
+    .wrap(100)
     .help()
-    .strict()
-    .parseAsync();
+    .fail(async (msg, err, _yargs) => {
```
Comment (hojarasca) [inline cli/src/lib/cli.ts:175]: Custom error handling to do a better job communicating known errors.

File: cli/src/lib/env-builder.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
+import { BlockExplorerClient } from "./block-explorer-client.js";
+import type { Network } from "./constants.js";
+import { GithubClient } from "./github-client.js";
+import { RpcClient } from "./rpc-client.js";
+
+export class EnvBuilder {
```
Comment (hojarasca) [inline cli/src/lib/env-builder.ts:6]: This class holds all the configuration and builds the clients to interact with other systems.

File: cli/src/lib/errors.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
+import * as console from "node:console";
+import type { Network } from "./constants.js";
+
```
Comment (hojarasca) [inline cli/src/lib/errors.ts:3]: Custom errors to display proper messages

File: cli/src/lib/importer.ts [hunk 7]
```diff
@@ -1,54 +1,17 @@
...
-  } catch (e) {
-    if (e instanceof ZodError) {
-      process.env.DEBUG === "1" && console.error(e.message);
-    } else {
-      console.error(e);
-    }
-    // TODO: Add a logging system and add debug logs for parse failure
-    return { valid: false };
-  }
-};
```
Comment (hojarasca) [inline cli/src/lib/importer.ts:51]: All this red was dead code.

File: cli/src/lib/importer.ts [hunk 8]
```diff
@@ -70,16 +33,31 @@ function translateNetwork(n: Network) {
 }
 
 export const lookupAndParse = async (
```
Comment (hojarasca) [inline cli/src/lib/importer.ts:35]: Added validations and throwing more specific errors

File: cli/src/lib/rpc-client.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+  http,
+} from "viem";
+import type { TypeOf, ZodType } from "zod";
+
+const DEFAULT_URLS = {
+  mainnet: "https://ethereum-rpc.publicnode.com",
+  sepolia: "https://ethereum-sepolia-rpc.publicnode.com",
+};
+
+export class RpcClient {
```
Comment (hojarasca) [inline cli/src/lib/rpc-client.ts:17]: We used to have to functions to do rpc calls. I decided to create this class to be able to hold configuration regarding rpc calls.

Comment (hojarasca) [reply cli/src/lib/rpc-client.ts:17]: The old functions were inside "contract-read.ts" (that file was deleted) and they were called `contractRead` and `contractReadRaw`.

File: cli/src/lib/system-contract-change.ts [hunk 10]
```diff
@@ -17,7 +19,19 @@ export class SystemContractChange {
...
+      // Some system contracts do not have the code available in the block explorer. But this is not an error.
+      // For these contracts we cannot show the current source code.
+      if (e instanceof ContracNotVerified) {
+        const content = {
+          content: `Code for contract ${this.address} (${this.name}) is not available in block explorer`,
+        };
+        return new ContractData(this.name, { "messate.txt": content }, this.address);
+      }
+      throw e;
+    }
```
Comment (hojarasca) [inline cli/src/lib/system-contract-change.ts:34]: Another polemic error handling. Again, I believe this is correct, even when it sounds weird.

File: cli/src/lib/upgrade-changes.ts [hunk 11]
```diff
@@ -77,8 +77,8 @@ export class UpgradeChanges {
     const verifier = new VerifierContract(
       txFile.proposeUpgradeTx.verifier,
       txFile.proposeUpgradeTx.verifierParams.recursionCircuitsSetVksHash,
-      txFile.proposeUpgradeTx.verifierParams.recursionNodeLevelVkHash,
-      txFile.proposeUpgradeTx.verifierParams.recursionLeafLevelVkHash
+      txFile.proposeUpgradeTx.verifierParams.recursionLeafLevelVkHash,
+      txFile.proposeUpgradeTx.verifierParams.recursionNodeLevelVkHash
```
Comment (hojarasca) [inline cli/src/lib/upgrade-changes.ts:81]: bugfix, this is the right order.

File: cli/src/lib/block-explorer-client.ts (hunk 1 above)
Comment (HermanObst) [reply cli/src/lib/block-explorer-client.ts:132]: This means that we got an Etherscan response that we are not able to parse, and convert into a proper class?
In what case can this happen? Dont we would need to say something?

File: cli/src/lib/cli.ts [hunk 12]
```diff
@@ -3,19 +3,33 @@ import { hideBin } from "yargs/helpers";
 import { NetworkSchema } from ".";
```
[the remark this answers is not in the document]
Comment (HermanObst) [reply cli/src/lib/cli.ts:3]: Thanks for this

File: cli/src/lib/cli.ts (hunk 3 above)
Comment (HermanObst) [reply cli/src/lib/cli.ts:69]: Beautiful!

File: cli/src/lib/env-builder.ts [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    this.githubApiKey = maybeKey;
+  }
+
+  withRpcUrl(maybeUrl: string | undefined): void {
+    this.rpcUrl = maybeUrl;
+  }
+
+  // Get components
+
+  get network(): Network {
```
Comment (HermanObst) [inline cli/src/lib/env-builder.ts:36]: How can you have this class without the network? Dont we have a default one?

**Review by HermanObst — APPROVED** [review]

**Comment by HermanObst** [comment]
I still need to check in detail the integrations test. But we can move forward as it is.
Great job @hojarasca

**Pull request merged**